core/src/main/scala/org/http4s/util/Compression.scala

```diff
@@ -0,0 +1,100 @@
+package org.http4s.util
+
+import java.util.zip.{CRC32, Deflater}
+import javax.xml.bind.DatatypeConverter
+
+import cats.Functor
+import fs2.compress.{deflate, inflate}
+import fs2.{Chunk, Pipe, Pull, Pure, Segment, Stream}
+import fs2.Stream.{chunk => fs2chunk} // because of org.http4s.util.chunk
+import org.http4s.headers.{`Content-Encoding`, `Content-Length`, `Transfer-Encoding`}
+import org.http4s.{ContentCoding, Response, TransferCoding}
+import org.log4s.getLogger
+
+object Compression {
+
+  private[this] val logger = getLogger
+
+  def zipResponse[F[_]: Functor](bufferSize: Int, level: Int, resp: Response[F]): Response[F] = {
+    logger.trace("GZip middleware encoding content")
+    // Need to add the Gzip header and trailer
+    val trailerGen = new TrailerGen()
+    val b = fs2chunk(header) ++
+      resp.body
+        .through(trailer(trailerGen))
+        .through(
+          deflate(
+            level = level,
+            nowrap = true,
+            bufferSize = bufferSize
+          )) ++
+      fs2chunk(trailerFinish(trailerGen))
+    resp
+      .removeHeader(`Content-Length`)
+      .putHeaders(`Content-Encoding`(ContentCoding.gzip))
+      .copy(body = b)
+  }
+
+  def unzipResponse[F[_]: Functor](bufferSize: Int, resp: Response[F]): Response[F] = {
+    logger.trace("GZip middleware decoding content")
+    val b =
+      // TODO We should validate the header. Futhermore, there are
+      // optional extra headers that can extend the length before we
+      // get to the deflate-compressed body.
+      //
+      // We're also not validating the trailer.
```

Comment:
I wonder what we should do if we start decoding this and deem it to be invalid GZip.  I think we need either an error return type or restore the original stream.